File: forge/src/main/java/com/eliasnvx/krylix/forge/network/NetworkPackets.java
package com.eliasnvx.krylix.forge.network;

import com.eliasnvx.krylix.Krylix;
import com.eliasnvx.krylix.model.KillEntry;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraftforge.network.ChannelBuilder;
import net.minecraftforge.network.PacketDistributor;
import net.minecraftforge.network.SimpleChannel;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NetworkPackets {
    public static final SimpleChannel CHANNEL = ChannelBuilder.named(
        ResourceLocation.fromNamespaceAndPath(Krylix.MOD_ID, "main")
    ).networkProtocolVersion(1)
     .simpleChannel();

    public static void register() {
        CHANNEL.messageBuilder(KillNotificationPacket.class)
            .encoder(KillNotificationPacket::encode)
            .decoder(KillNotificationPacket::decode)
            .consumerMainThread((msg, ctx) -> msg.handleOnClient())
            .add();

        CHANNEL.messageBuilder(DeathRecapPacket.class)
            .encoder(DeathRecapPacket::encode)
            .decoder(DeathRecapPacket::decode)
            .consumerMainThread((msg, ctx) -> msg.handleOnClient())
            .add();

        CHANNEL.messageBuilder(MobStatsSyncPacket.class)
            .encoder(MobStatsSyncPacket::encode)
            .decoder(MobStatsSyncPacket::decode)
            .consumerMainThread((msg, ctx) -> msg.handleOnClient())
            .add();

        CHANNEL.messageBuilder(PlayerStatsSyncPacket.class)
            .encoder(PlayerStatsSyncPacket::encode)
            .decoder(PlayerStatsSyncPacket::decode)
            .consumerMainThread((msg, ctx) -> msg.handleOnClient())
            .add();
    }

    public static void sendToAll(Object packet) {
        CHANNEL.send(packet, PacketDistributor.ALL.noArg());
    }

    public static void sendToPlayer(ServerPlayer player, Object packet) {
        CHANNEL.send(packet, PacketDistributor.PLAYER.with(player));
    }

    public static void sendToDimension(ServerLevel dimension, Object packet) {
        CHANNEL.send(packet, PacketDistributor.DIMENSION.with(dimension.dimension()));
    }

    // Packet Classes

    public record KillNotificationPacket(
        String killerName,
        String killerUUIDString,
        String victimName,
        String victimUUIDString,
        float killerHealth,
        String weaponName,
        Double distance,
        long timestamp,
        boolean isHeadshot,
        boolean isSmash,
        boolean isCritical
    ) {
        public void encode(FriendlyByteBuf buf) {
            buf.writeUtf(killerName != null ? killerName : "");
            buf.writeUtf(killerUUIDString != null ? killerUUIDString : "");
            buf.writeUtf(victimName);
            buf.writeUtf(victimUUIDString != null ? victimUUIDString : "");
            buf.writeFloat(killerHealth);
            buf.writeUtf(weaponName);
            buf.writeDouble(distance != null ? distance : 0.0);
            buf.writeLong(timestamp);
            buf.writeBoolean(isHeadshot);
            buf.writeBoolean(isSmash);
            buf.writeBoolean(isCritical);
        }

        public static KillNotificationPacket decode(FriendlyByteBuf buf) {
            String killerName = buf.readUtf();
            if (killerName.isEmpty()) killerName = null;
            String killerUUID = buf.readUtf();
            if (killerUUID.isEmpty()) killerUUID = null;
            String victimName = buf.readUtf();
            String victimUUID = buf.readUtf();
            if (victimUUID.isEmpty()) victimUUID = null;
            float killerHealth = buf.readFloat();
            String weaponName = buf.readUtf();
            double d = buf.readDouble();
            Double distance = d == 0.0 ? null : d;
            long timestamp = buf.readLong();
            boolean isHeadshot = buf.readBoolean();
            boolean isSmash = buf.readBoolean();
            boolean isCritical = buf.readBoolean();

            return new KillNotificationPacket(
                killerName, killerUUID, victimName, victimUUID, killerHealth, weaponName, distance, timestamp, isHeadshot, isSmash, isCritical
            );
        }

        public void handleOnClient() {
            KillEntry entry = new KillEntry(
                killerName, killerUUIDString, victimName, victimUUIDString, killerHealth, weaponName, distance, timestamp, isHeadshot, isSmash, isCritical
            );
            com.eliasnvx.krylix.forge.client.KillFeedHud.addEntry(entry);
        }
    }

    public record DeathRecapPacket(
        String killerName,
        String killerUUIDString,
        float killerHealth,
        float killerMaxHealth,
        String weaponName,
        Double distance,
        float damageDealtToKiller,
        boolean isSmash,
        boolean isCritical,
        boolean isLongshot
    ) {
        public void encode(FriendlyByteBuf buf) {
            buf.writeUtf(killerName);
            buf.writeUtf(killerUUIDString != null ? killerUUIDString : "");
            buf.writeFloat(killerHealth);
            buf.writeFloat(killerMaxHealth);
            buf.writeUtf(weaponName);
            buf.writeDouble(distance != null ? distance : 0.0);
            buf.writeFloat(damageDealtToKiller);
            buf.writeBoolean(isSmash);
            buf.writeBoolean(isCritical);
            buf.writeBoolean(isLongshot);
        }

        public static DeathRecapPacket decode(FriendlyByteBuf buf) {
            String killerName = buf.readUtf();
            String killerUUID = buf.readUtf();
            if (killerUUID.isEmpty()) killerUUID = null;
            float killerHealth = buf.readFloat();
            float killerMaxHealth = buf.readFloat();
            String weaponName = buf.readUtf();
            double d = buf.readDouble();
            Double distance = d == 0.0 ? null : d;
            float damageDealt = buf.readFloat();
            boolean isSmash = buf.readBoolean();
            boolean isCritical = buf.readBoolean();
            boolean isLongshot = buf.readBoolean();

            return new DeathRecapPacket(killerName, killerUUID, killerHealth, killerMaxHealth, weaponName, distance, damageDealt, isSmash, isCritical, isLongshot);
        }

        public void handleOnClient() {
            com.eliasnvx.krylix.forge.client.DeathRecapClient.setRecap(this);
        }
    }

    public record MobStatsSyncPacket(Map<String, Integer> kills) {
        public void encode(FriendlyByteBuf buf) {
            buf.writeVarInt(kills.size());
            for (Map.Entry<String, Integer> entry : kills.entrySet()) {
                buf.writeUtf(entry.getKey());
                buf.writeVarInt(entry.getValue());
            }
        }

        public static MobStatsSyncPacket decode(FriendlyByteBuf buf) {
            int size = buf.readVarInt();
            Map<String, Integer> kills = new HashMap<>();
            for (int i = 0; i < size; i++) {
                kills.put(buf.readUtf(), buf.readVarInt());
            }
            return new MobStatsSyncPacket(kills);
        }

        public void handleOnClient() {
            com.eliasnvx.krylix.forge.client.MobStatsClient.updateStats(kills);
        }
    }

    public record PlayerStatEntry(String uuid, String name, int kills, int deaths, int mobKills) {}

    public record PlayerStatsSyncPacket(List<PlayerStatEntry> entries) {
        public void encode(FriendlyByteBuf buf) {
            buf.writeVarInt(entries.size());
            for (PlayerStatEntry entry : entries) {
                buf.writeUtf(entry.uuid());
                buf.writeUtf(entry.name());
                buf.writeVarInt(entry.kills());
                buf.writeVarInt(entry.deaths());
                buf.writeVarInt(entry.mobKills());
            }
        }

        public static PlayerStatsSyncPacket decode(FriendlyByteBuf buf) {
            int size = buf.readVarInt();
            List<PlayerStatEntry> entries = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                entries.add(new PlayerStatEntry(buf.readUtf(), buf.readUtf(), buf.readVarInt(), buf.readVarInt(), buf.readVarInt()));
            }
            return new PlayerStatsSyncPacket(entries);
        }

        public void handleOnClient() {
            com.eliasnvx.krylix.forge.client.PlayerStatsClient.updateStats(entries);
        }
    }
}

File: forge/src/main/kotlin/com/eliasnvx/krylix/forge/network/NetworkPackets.kt
package com.eliasnvx.krylix.forge.network

import com.eliasnvx.krylix.Krylix
import com.eliasnvx.krylix.forge.PlayerStat
import com.eliasnvx.krylix.model.KillEntry
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerPlayer
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn
import net.minecraftforge.network.NetworkDirection
import net.minecraftforge.network.NetworkEvent
import net.minecraftforge.network.NetworkRegistry
import net.minecraftforge.network.PacketDistributor
import net.minecraftforge.network.simple.SimpleChannel
import java.util.function.Supplier

/**
 * Network packets для синхронизации kill feed между сервером и клиентами
 */
object NetworkPackets {
    private const val PROTOCOL_VERSION = "1"

    val CHANNEL: SimpleChannel = NetworkRegistry.newSimpleChannel(
        ResourceLocation(Krylix.MOD_ID, "main"),
        { PROTOCOL_VERSION },
        { PROTOCOL_VERSION == it },
        { PROTOCOL_VERSION == it }
    )

    init {
        CHANNEL.messageBuilder(KillNotificationPacket::class.java, 0)
            .encoder(KillNotificationPacket::encode)
            .decoder(KillNotificationPacket::decode)
            .consumerMainThread { msg, ctx ->
                msg.handle(ctx)
                true
            }
            .add()

        CHANNEL.messageBuilder(MobStatsSyncPacket::class.java, 1)
            .encoder(MobStatsSyncPacket::encode)
            .decoder(MobStatsSyncPacket::decode)
            .consumerMainThread { msg, ctx ->
                msg.handle(ctx)
                true
            }
            .add()

        CHANNEL.messageBuilder(PlayerStatsSyncPacket::class.java, 2)
            .encoder(PlayerStatsSyncPacket::encode)
            .decoder(PlayerStatsSyncPacket::decode)
            .consumerMainThread { msg, ctx ->
                msg.handle(ctx)
                true
            }
            .add()
    }

    /**
     * Отправляет уведомление об убийстве всем игрокам на сервере
     */
    fun sendKillNotificationToAll(killEntry: KillEntry) {
        val packet = KillNotificationPacket(killEntry)
        CHANNEL.send(PacketDistributor.ALL.noArg(), packet)
        Krylix.LOGGER.debug("Sent kill notification to all clients: $killEntry")
    }

    /**
     * Отправляет уведомление конкретному игроку
     */
    fun sendKillNotificationToPlayer(player: ServerPlayer, killEntry: KillEntry) {
        val packet = KillNotificationPacket(killEntry)
        CHANNEL.send(PacketDistributor.PLAYER.with { player }, packet)
        Krylix.LOGGER.debug("Sent kill notification to ${player.name.string}: $killEntry")
    }

    /**
     * Отправляет уведомление об убийстве только перечисленным игрокам (напр. в пределах одного измерения)
     */
    fun sendKillNotificationToPlayers(players: Collection<ServerPlayer>, killEntry: KillEntry) {
        val packet = KillNotificationPacket(killEntry)
        players.forEach { player -> CHANNEL.send(PacketDistributor.PLAYER.with { player }, packet) }
        Krylix.LOGGER.debug("Sent kill notification to ${players.size} players: $killEntry")
    }

    /**
     * Отправляет снапшот статистики убийств мобов конкретному игроку
     */
    fun sendMobStatsSync(player: ServerPlayer, kills: Map<String, Int>) {
        CHANNEL.send(PacketDistributor.PLAYER.with { player }, MobStatsSyncPacket(kills))
        Krylix.LOGGER.debug("Sent mob stats sync to ${player.name.string}: $kills")
    }

    /**
     * Отправляет снапшот PvP-рейтинга конкретному игроку (напр. при входе)
     */
    fun sendPlayerStatsSync(player: ServerPlayer, stats: Map<String, PlayerStat>) {
        CHANNEL.send(PacketDistributor.PLAYER.with { player }, PlayerStatsSyncPacket(toEntries(stats)))
    }

    /**
     * Рассылает снапшот PvP-рейтинга всем игрокам (после каждого PvP-килла) — рейтинг это
     * общесерверная сводка, поэтому рассылается всем независимо от restrictBroadcastToSameDimension
     */
    fun broadcastPlayerStatsSync(stats: Map<String, PlayerStat>) {
        CHANNEL.send(PacketDistributor.ALL.noArg(), PlayerStatsSyncPacket(toEntries(stats)))
    }

    private fun toEntries(stats: Map<String, PlayerStat>): List<PlayerStatsSyncPacket.PlayerStatEntry> =
        stats.map { (uuid, stat) -> PlayerStatsSyncPacket.PlayerStatEntry(uuid, stat.lastName, stat.kills, stat.deaths, stat.mobKills) }

    /**
     * Packet для уведомления об убийстве
     */
    data class KillNotificationPacket(
        val killerName: String?,
        val victimName: String,
        val weaponName: String,
        val distance: Double?,
        val timestamp: Long,
        val isHeadshot: Boolean
    ) {

        constructor(killEntry: KillEntry) : this(
            killEntry.killerName,
            killEntry.victimName,
            killEntry.weaponName,
            killEntry.distance,
            killEntry.timestamp,
            killEntry.isHeadshot
        )

        fun encode(buf: FriendlyByteBuf) {
            buf.writeUtf(killerName ?: "")
            buf.writeUtf(victimName)
            buf.writeUtf(weaponName)
            buf.writeDouble(distance ?: 0.0)
            buf.writeLong(timestamp)
            buf.writeBoolean(isHeadshot)
        }

        fun handle(ctx: Supplier<NetworkEvent.Context>) {
            ctx.get().enqueueWork {
                // Обрабатываем только на клиенте
                if (ctx.get().direction == NetworkDirection.PLAY_TO_CLIENT) {
                    handleOnClient()
                }
            }
            ctx.get().packetHandled = true
        }

        @OnlyIn(Dist.CLIENT)
        private fun handleOnClient() {
            val killEntry = KillEntry(
                killerName = killerName,
                victimName = victimName,
                weaponName = weaponName,
                distance = distance,
                timestamp = timestamp,
                isHeadshot = isHeadshot
            )

            // Добавляем в client-side HUD только на клиенте
            com.eliasnvx.krylix.forge.client.KillFeedHud.addNotification(killEntry)

            Krylix.LOGGER.debug("Received kill notification on client: $killEntry")
        }

        companion object {
            fun decode(buf: FriendlyByteBuf): KillNotificationPacket {
                val killerName = buf.readUtf().takeIf { it.isNotEmpty() }
                val victimName = buf.readUtf()
                val weaponName = buf.readUtf()
                val distance = buf.readDouble().takeIf { it != 0.0 }
                val timestamp = buf.readLong()
                val isHeadshot = buf.readBoolean()

                return KillNotificationPacket(
                    killerName,
                    victimName,
                    weaponName,
                    distance,
                    timestamp,
                    isHeadshot
                )
            }
        }
    }

    /**
     * Packet со снапшотом статистики убийств мобов (registry id -> количество)
     */
    data class MobStatsSyncPacket(val kills: Map<String, Int>) {

        fun encode(buf: FriendlyByteBuf) {
            buf.writeVarInt(kills.size)
            kills.forEach { (id, count) ->
                buf.writeUtf(id)
                buf.writeVarInt(count)
            }
        }

        fun handle(ctx: Supplier<NetworkEvent.Context>) {
            ctx.get().enqueueWork {
                if (ctx.get().direction == NetworkDirection.PLAY_TO_CLIENT) {
                    handleOnClient()
                }
            }
            ctx.get().packetHandled = true
        }

        @OnlyIn(Dist.CLIENT)
        private fun handleOnClient() {
            com.eliasnvx.krylix.forge.client.MobStatsClient.updateStats(kills)
        }

        companion object {
            fun decode(buf: FriendlyByteBuf): MobStatsSyncPacket {
                val size = buf.readVarInt()
                val kills = buildMap {
                    repeat(size) {
                        val id = buf.readUtf()
                        val count = buf.readVarInt()
                        put(id, count)
                    }
                }
                return MobStatsSyncPacket(kills)
            }
        }
    }

    /**
     * Packet со снапшотом PvP-рейтинга игроков
     */
    data class PlayerStatsSyncPacket(val entries: List<PlayerStatEntry>) {

        data class PlayerStatEntry(val uuid: String, val name: String, val kills: Int, val deaths: Int, val mobKills: Int)

        fun encode(buf: FriendlyByteBuf) {
            buf.writeVarInt(entries.size)
            entries.forEach { entry ->
                buf.writeUtf(entry.uuid)
                buf.writeUtf(entry.name)
                buf.writeVarInt(entry.kills)
                buf.writeVarInt(entry.deaths)
                buf.writeVarInt(entry.mobKills)
            }
        }

        fun handle(ctx: Supplier<NetworkEvent.Context>) {
            ctx.get().enqueueWork {
                if (ctx.get().direction == NetworkDirection.PLAY_TO_CLIENT) {
                    handleOnClient()
                }
            }
            ctx.get().packetHandled = true
        }

        @OnlyIn(Dist.CLIENT)
        private fun handleOnClient() {
            com.eliasnvx.krylix.forge.client.PlayerStatsClient.updateStats(entries)
        }

        companion object {
            fun decode(buf: FriendlyByteBuf): PlayerStatsSyncPacket {
                val size = buf.readVarInt()
                val entries = buildList {
                    repeat(size) {
                        val uuid = buf.readUtf()
                        val name = buf.readUtf()
                        val kills = buf.readVarInt()
                        val deaths = buf.readVarInt()
                        val mobKills = buf.readVarInt()
                        add(PlayerStatEntry(uuid, name, kills, deaths, mobKills))
                    }
                }
                return PlayerStatsSyncPacket(entries)
            }
        }
    }
}
